// @alchemy/aa-core/dist/esm/provider/schema.js
import { Address } from "abitype/zod";
import z from "zod";
import { createPublicErc4337ClientSchema } from "../client/schema.js";
import { BigNumberishRangeSchema, ChainSchema, PercentageSchema, } from "../utils/index.js";
export const UserOperationFeeOptionsFieldSchema = BigNumberishRangeSchema.merge(PercentageSchema).partial();
export const UserOperationFeeOptionsSchema = z
    .object({
    maxFeePerGas: UserOperationFeeOptionsFieldSchema,
    maxPriorityFeePerGas: UserOperationFeeOptionsFieldSchema,
    callGasLimit: UserOperationFeeOptionsFieldSchema,
    verificationGasLimit: UserOperationFeeOptionsFieldSchema,
    preVerificationGas: UserOperationFeeOptionsFieldSchema,
})
    .partial()
    .strict();
export const SmartAccountProviderOptsSchema = z
    .object({
    txMaxRetries: z.number().min(0).optional().default(5),
    txRetryIntervalMs: z.number().min(0).optional().default(2000),
    txRetryMulitplier: z.number().min(0).optional().default(1.5),
    feeOptions: UserOperationFeeOptionsSchema.optional(),
})
    .strict();
export const createSmartAccountProviderConfigSchema = () => {
    return z.object({
        rpcProvider: z.union([
            z.string(),
            createPublicErc4337ClientSchema(),
        ]),
        chain: ChainSchema,
        entryPointAddress: Address.optional(),
        opts: SmartAccountProviderOptsSchema.optional().default(SmartAccountProviderOptsSchema.parse({})),
    });
};
